crates/astria-proto/src/native/sequencer/v1alpha1.rs

```diff
@@ -3,18 +3,720 @@ use std::{
+        }
+    }
+}
+
+impl SequenceActionError {
+    fn chain_id_len(inner: IncorrectChainIdLength) -> Self {
+        Self {
+            kind: SequenceActionErrorKind::ChainId(inner),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum SequenceActionErrorKind {
+    ChainId(IncorrectChainIdLength),
+}
+
+#[derive(Clone, Debug)]
+pub struct TransferAction {
+    pub to: Address,
+    pub amount: u128,
+}
+
+impl TransferAction {
+    #[must_use]
+    pub fn into_raw(self) -> raw::TransferAction {
+        let Self {
+            to,
+            amount,
+        } = self;
+        raw::TransferAction {
+            to: to.to_vec(),
+            amount: Some(amount.into()),
+        }
+    }
+
+    #[must_use]
+    pub fn to_raw(&self) -> raw::TransferAction {
+        let Self {
+            to,
+            amount,
+        } = self;
+        raw::TransferAction {
+            to: to.to_vec(),
+            amount: Some((*amount).into()),
+        }
+    }
+
+    /// Convert from a raw, unchecked protobuf [`raw::TransferAction`].
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the raw action's `to` address did not have the expected
+    /// length.
+    pub fn try_from_raw(proto: raw::TransferAction) -> Result<Self, TransferActionError> {
+        let raw::TransferAction {
+            to,
+            amount,
+        } = proto;
+        let to = Address::try_from_slice(&to).map_err(TransferActionError::address)?;
+        let amount = amount.map_or(0, Into::into);
+        Ok(Self {
+            to,
+            amount,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct TransferActionError {
+    kind: TransferActionErrorKind,
+}
+
+impl TransferActionError {
+    fn address(inner: IncorrectAddressLength) -> Self {
+        Self {
+            kind: TransferActionErrorKind::Address(inner),
+        }
+    }
+}
+
+impl Display for TransferActionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.kind {
+            TransferActionErrorKind::Address(_) => {
+                f.pad("`to` field did not contain a valid address")
+            }
+        }
+    }
+}
+
+impl Error for TransferActionError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match &self.kind {
+            TransferActionErrorKind::Address(e) => Some(e),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum TransferActionErrorKind {
+    Address(IncorrectAddressLength),
+}
+
+/// Indicates that the protobuf response contained an array field that was not 20 bytes long.
+#[derive(Debug)]
+pub struct IncorrectChainIdLength {
+    received: usize,
+}
+
+impl Display for IncorrectChainIdLength {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "expected 32 bytes, got {}", self.received)
+    }
+}
+
+impl Error for IncorrectChainIdLength {}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ChainId(pub [u8; CHAIN_ID_LEN]);
```

Comment:
I am a bit torn on using a newtype wrapper here. It makes constructing chain IDs more straightforward by using `ChainId::with_hashed_bytes("my_test_chain")`, but at the expense of introducing yet another type. So the signature of `sequence_action` could instead be:
```rust
struct SequenceAction {
    chain_id: [u8; 32],
    data: Vec<u8>,
}
```
which is a bit more obvious. Then we'd provide utilities like `chain_id::with_hashed_bytes -> [u8; 32]` and `chain_id::try_from_slice -> [u8; 32]`, which are also a biy ugly.

Reply:
I think a newtype is fine, but idk if I like the `with_hashed_bytes`, since we now have a plaintext string we input, then hash to a `chain_id`, then hash again for a `Namespace`, which seems overkill. would prefer just having a `ChainId::new()` which takes some bytes and pads them to 32 bytes if needed